src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -750,34 +821,59 @@ private DiagnosticData UpdatePosition(DiagnosticData diagnostic, SyntaxTree tree
                 helpLink: diagnostic.HelpLink);
         }
 
-        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
+        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, SyntaxTree tree, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
         {
-            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, span)).Select(d => DiagnosticData.Create(document, d)) : null;
+            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, tree, span)).Select(d => DiagnosticData.Create(document, d)) : null;
```

Comment:
So now we handle diagnostics reported in end actions with tree locations, but we still don't handle non-local diagnostics reported in different documents by document actions (symbol/syntax/semantic model actions)? I agree that's a less of a concern then the issue addressed by this change, but @JohnHamby already mentioned few analyzer examples in our internal discussions where such an analyzer is pretty likely. We should atleast open a tracking github issue so that we address it eventually (v1 or v2 engine)